Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs

```diff
@@ -298,14 +307,30 @@ private static void PasteAction(SerializedProperty arrayProperty, string jsonToI
         private static int PasteBindingOrComposite(SerializedProperty arrayProperty, string json, int index, string actionName, bool createCompositeParts = true)
         {
             var pastePartOfComposite = IsPartOfComposite(json);
+            bool currentPartOfComposite = false;
+            bool currentIsComposite = false;
+
             if (index > 0)
             {
                 var currentProperty = arrayProperty.GetArrayElementAtIndex(index - 1);
-                var currentIsComposite = IsComposite(currentProperty) || IsPartOfComposite(currentProperty);
+                currentPartOfComposite = IsPartOfComposite(currentProperty);
+                currentIsComposite = IsComposite(currentProperty) || currentPartOfComposite;
                 if (pastePartOfComposite && !currentIsComposite) //prevent pasting part of composite into non-composite
                     return index;
             }
-            index = pastePartOfComposite || s_State.selectionType == SelectionType.Action ? index : Selectors.GetSelectedBindingIndexAfterCompositeBindings(s_State) + 1;
+
+            // Update the target index for special cases
+            if (pastePartOfComposite && !currentPartOfComposite)
+            {
+                // Pasting into a Composite and CompositePart isn't the target, i.e. Composite "root" selected, paste at the end of the composite
+                index = Selectors.GetSelectedBindingIndexAfterCompositeBindings(s_State) + 1;
+            }
+            else if (!pastePartOfComposite && s_State.selectionType != SelectionType.Action)
+            {
+                // Pasting with a Binding selected needs to skip all the CompositeParts (if any)
+                index = Selectors.GetSelectedBindingIndexAfterCompositeBindings(s_State) + 1;
```

Comment:
the if/else have identical content. maybe declare boolean variables before and check them into one combined if statement?

Reply:
The logic was kinda confusing for me, and so I did it this way because I found it easier to reason through, but I do agree it's not optimal. I'll see if can rework it so it's not duplicating checks but still not too confusing.

Reply:
How about if you take the content of the if statements into two booleans with comments?

Reply:
Yeah that's a good suggestion, and I think I'll go that way.

Reply:
Reworked the conditions to remove duplicated line, and also fix regression pasting Composite: 630fe8d551ecfe45a0245c77583f76b8faf1f285